src/interpreter.js

```diff
@@ -1,58 +1,17 @@
-const { getFieldInfo, tryCatch } = require('./utils')
+const { getFieldInfo, isFieldInfo, tryCatch } = require('./utils')
 const reduce = require('lodash.reduce')
 const get = require('lodash.get')
 const Validator = require('protodef-validator')
+const defaultDatatypes = require('./datatypes/interpreter')
 
-function isFieldInfo (type) {
-  return typeof type === 'string' ||
-    (Array.isArray(type) && typeof type[0] === 'string') ||
-    type.type
-}
-
-function findArgs (acc, v, k) {
-  if (typeof v === 'string' && v.charAt(0) === '$') { acc.push({ 'path': k, 'val': v.substr(1) }) } else if (Array.isArray(v) || typeof v === 'object') { acc = acc.concat(reduce(v, findArgs, []).map((v) => ({ 'path': k + '.' + v.path, 'val': v.val }))) }
-  return acc
-}
-
-function setField (path, val, into) {
-  const c = path.split('.').reverse()
-  while (c.length > 1) {
-    into = into[c.pop()]
-  }
-  into[c.pop()] = val
-}
-
-function extendType (functions, defaultTypeArgs) {
-  const json = JSON.stringify(defaultTypeArgs)
-  const argPos = reduce(defaultTypeArgs, findArgs, [])
-  function produceArgs (typeArgs) {
-    const args = JSON.parse(json)
-    argPos.forEach((v) => {
-      setField(v.path, typeArgs[v.val], args)
-    })
-    return args
-  }
-  return [function read (buffer, offset, typeArgs, context) {
-    return functions[0].call(this, buffer, offset, produceArgs(typeArgs), context)
-  }, function write (value, buffer, offset, typeArgs, context) {
-    return functions[1].call(this, value, buffer, offset, produceArgs(typeArgs), context)
-  }, function sizeOf (value, typeArgs, context) {
-    if (typeof functions[2] === 'function') { return functions[2].call(this, value, produceArgs(typeArgs), context) } else { return functions[2] }
-  }]
-}
+// FIXME: Increases performance but introduces unexpected bugs in future
```

Comment:
Can you elaborate on this comment ? What kind of unexpected bugs should we worry about ?

Reply:
https://github.com/ProtoDef-io/node-protodef/pull/107/files/43f7401c0be36b6d001db2c18a24ad8d7c91b34a#diff-14cffeee7a34687f7a373b01152975d8R135
Before this PR the interpreter every time does copy of datatype and merges it with passed arguments, I don't know why this behavior was there. Maybe @rom1504 should explain that to me.

Reply:
That kind of bugs:
https://github.com/PrismarineJS/node-minecraft-protocol/pull/731/checks?check_run_id=777137891#step:6:239

Reply:
Switched from `JSON.stringify` to `_.cloneDeep`, it was fast enough. Also fixed `defaultTypeArgs !== 'object'` to `typeof defaultTypeArgs !== 'object'` (how did I miss that before?)